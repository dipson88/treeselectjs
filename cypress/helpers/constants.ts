export const titleSelectors = {
  EnglandGroup: '[title="England"]',
  LondonGroup: '[title="London"]',
  ChelseaItem: '[title="Chelsea"]',
  WestEndItem: '[title="West End"]',
  BrightonItem: '[title="Brighton"]',
  FranceGroup: '[title="France"]',
  ParisItem: '[title="Paris"]',
  LyonItem: '[title="Lyon"]'
}

export const classes = {
  parent: 'treeselect',
  input: {
    arrow: 'treeselect-input__arrow',
    base: 'treeselect-input',
    bottom: 'treeselect-input--bottom',
    clear: 'treeselect-input__clear',
    focused: 'treeselect-input--focused',
    opened: 'treeselect-input--opened',
    tagsElement: 'treeselect-input__tags-element',
    top: 'treeselect-input--top'
  },
  list: {
    base: 'treeselect-list',
    bottom: 'treeselect-list--bottom',
    itemArrow: 'treeselect-list__item-icon',
    itemChecked: 'treeselect-list__item--checked',
    itemDisabled: 'treeselect-list__item--disabled',
    itemFocused: 'treeselect-list__item--focused',
    itemGroup: 'treeselect-list__item--group',
    itemHidden: 'treeselect-list__item--hidden',
    itemPartialChecked: 'treeselect-list__item--partial-checked',
    top: 'treeselect-list--top'
  }
}

export const classesSelectors = {
  parent: `.${classes.parent}`,
  input: {
    arrow: `.${classes.input.arrow}`,
    base: `.${classes.input.base}`,
    bottom: `.${classes.input.bottom}`,
    clear: `.${classes.input.clear}`,
    focused: `.${classes.input.focused}`,
    opened: `.${classes.input.opened}`,
    tagsElement: `.${classes.input.tagsElement}`,
    top: `.${classes.input.top}`
  },
  list: {
    base: `.${classes.list.base}`,
    bottom: `.${classes.list.bottom}`,
    itemArrow: `.${classes.list.itemArrow}`,
    itemChecked: `.${classes.list.itemChecked}`,
    itemDisabled: `.${classes.list.itemDisabled}`,
    itemFocused: `.${classes.list.itemFocused}`,
    itemGroup: `.${classes.list.itemGroup}`,
    itemHidden: `.${classes.list.itemHidden}`,
    itemPartialChecked: `.${classes.list.itemPartialChecked}`,
    top: `.${classes.list.top}`
  }
}
