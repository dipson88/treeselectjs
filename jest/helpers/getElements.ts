export const getTagsElements = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelectorAll('.treeselect-input__tags-element') as NodeListOf<HTMLElement>

export const getTagsElement = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelector('.treeselect-input__tags') as HTMLElement

export const getEditElement = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelector('.treeselect-input__edit') as HTMLElement

export const getArrowElement = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelector('.treeselect-input__arrow') as HTMLElement

export const getClearElement = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelector('.treeselect-input__clear') as HTMLElement

export const getListItems = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelectorAll('.treeselect-list__item') as NodeListOf<HTMLElement>

export const getNoResultsElement = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelector('.treeselect-list__empty') as HTMLElement

export const getListGroupsItems = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelectorAll('.treeselect-list__item--group') as NodeListOf<HTMLElement>

export const getListElement = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelector('.treeselect-list') as HTMLElement

export const getListSlotElement = (parentHtmlContainer: HTMLElement) =>
  parentHtmlContainer.querySelector('.treeselect-list__slot') as HTMLElement
