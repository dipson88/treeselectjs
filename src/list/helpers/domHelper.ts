import { type ValueOptionType, type IconsType } from '../../treeselectTypes'
import { type OptionsTreeMap, type TreeItem } from '../listTypes'
import { appendIconToElement } from '../../svgIcons'

export const updateDOM = ({
  optionsTreeMap,
  emptyListHtmlElement,
  iconElements,
  previousSingleSelectedValue,
  rtl
}: {
  optionsTreeMap: OptionsTreeMap
  emptyListHtmlElement: HTMLElement | null
  iconElements: IconsType
  previousSingleSelectedValue: ValueOptionType[]
  rtl: boolean
}) => {
  optionsTreeMap.forEach((option) => {
    const input = option.checkboxHtmlElement

    if (input) {
      input.checked = option.checked
    }

    updateCheckedClass({ option, previousSingleSelectedValue })
    updatePartialCheckedClass(option)
    updateDisabledCheckedClass(option)
    updateClosedClass({ option, iconElements })
    updateHiddenClass(option)
    updateLeftPaddingItems({ option, optionsTreeMap, rtl })
    updateCheckboxClass({ option, iconElements })
    updateGroupSelectableClass(option)
  })

  updateEmptyListClass({ optionsTreeMap, emptyListHtmlElement })
}

const updateLeftPaddingItems = ({
  option,
  optionsTreeMap,
  rtl
}: {
  option: TreeItem
  optionsTreeMap: OptionsTreeMap
  rtl: boolean
}) => {
  const isZeroLevel = option.level === 0
  const defaultPadding = 20
  const zeroLevelItemPadding = 5
  let padding = '0'

  if (isZeroLevel) {
    let isGroupsExistOnLevel = false

    for (const [_, item] of optionsTreeMap) {
      if (item.isGroup && item.level === option.level) {
        isGroupsExistOnLevel = true
        break
      }
    }

    const itemPadding = !option.isGroup && isGroupsExistOnLevel ? `${defaultPadding}px` : `${zeroLevelItemPadding}px`
    padding = option.isGroup ? '0' : itemPadding
  } else {
    padding = option.isGroup
      ? `${option.level * defaultPadding}px`
      : `${option.level * defaultPadding + defaultPadding}px`
  }

  const listItem = option.itemHtmlElement

  if (listItem) {
    if (rtl) {
      listItem.style.paddingRight = padding
    } else {
      listItem.style.paddingLeft = padding
    }

    // We can use css selectors to reset params with !important
    listItem.setAttribute('level', option.level.toString())
    listItem.setAttribute('group', option.isGroup.toString())
  }
}

const updateEmptyListClass = ({
  optionsTreeMap,
  emptyListHtmlElement
}: {
  optionsTreeMap: OptionsTreeMap
  emptyListHtmlElement: HTMLElement | null
}) => {
  let isNotEmpty = false

  for (const [_, option] of optionsTreeMap) {
    if (!option.hidden) {
      isNotEmpty = true
      break
    }
  }

  emptyListHtmlElement?.classList.toggle('treeselect-list__empty--hidden', isNotEmpty)
}

export const setAttributesFromHtmlAttr = (itemElement: HTMLDivElement, htmlAttr?: object) => {
  if (!htmlAttr) {
    return
  }

  Object.keys(htmlAttr).forEach((key) => {
    const value = htmlAttr[key as keyof object]

    if (typeof value === 'string') {
      itemElement.setAttribute(key, value)
    }
  })
}

const updateCheckedClass = ({
  option,
  previousSingleSelectedValue
}: {
  option: TreeItem
  previousSingleSelectedValue: ValueOptionType[]
}) => {
  const listItem = option.itemHtmlElement
  listItem?.classList.toggle('treeselect-list__item--checked', option.checked)
  const isCheckSingleSelected =
    Array.isArray(previousSingleSelectedValue) && previousSingleSelectedValue[0] === option.id && !option.disabled
  listItem?.classList.toggle('treeselect-list__item--single-selected', isCheckSingleSelected)
}

const updatePartialCheckedClass = (option: TreeItem) => {
  const listItem = option.itemHtmlElement
  listItem?.classList.toggle('treeselect-list__item--partial-checked', option.isPartialChecked)
}

const updateDisabledCheckedClass = (option: TreeItem) => {
  const listItem = option.itemHtmlElement
  listItem?.classList.toggle('treeselect-list__item--disabled', option.disabled)
}

const updateClosedClass = ({ option, iconElements }: { option: TreeItem; iconElements: IconsType }) => {
  const arrowIcon = option.arrowItemHtmlElement

  if (option.isGroup && arrowIcon) {
    const iconInnerElement = option.isClosed ? iconElements.arrowRight : iconElements.arrowDown
    appendIconToElement(iconInnerElement, arrowIcon)

    const listItem = option.itemHtmlElement
    listItem?.classList.toggle('treeselect-list__item--closed', option.isClosed)
  }
}

const updateHiddenClass = (option: TreeItem) => {
  const listItem = option.itemHtmlElement
  listItem?.classList.toggle('treeselect-list__item--hidden', option.hidden)
}

const updateCheckboxClass = ({ option, iconElements }: { option: TreeItem; iconElements: IconsType }) => {
  const icon = option.checkboxIconHtmlElement

  if (icon) {
    if (option.checked) {
      appendIconToElement(iconElements.check, icon)
    } else if (option.isPartialChecked) {
      appendIconToElement(iconElements.partialCheck, icon)
    } else {
      icon.innerHTML = ''
    }
  }
}

const updateGroupSelectableClass = (option: TreeItem) => {
  const listItem = option.itemHtmlElement
  listItem?.classList.toggle('treeselect-list__item--non-selectable-group', !option.isGroupSelectable)
}
