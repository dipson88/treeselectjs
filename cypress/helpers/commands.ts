import { classesSelectors } from './constants'

const { parent: parentSelector, input: inputSelector, list: listSelector } = classesSelectors

export const visitCypressDefaultPage = () => {
  cy.visit('http://localhost:5173/app/cypress/cypress-default.html')
}

export const treeselectClick = () => {
  cy.get(parentSelector).click()
}

export const treeselectFocus = () => {
  cy.get(parentSelector).find('input').focus()
}

export const treeselectType = (query: string) => {
  cy.get(parentSelector).type(query)
}

export const inputArrowClick = () => {
  cy.get(inputSelector.arrow).click()
}

export const outsideClick = () => {
  cy.get('body').click()
}

export const clearClick = () => {
  cy.get(inputSelector.clear).click()
}

export const expandAllGroups = () => {
  cy.get(listSelector.itemGroup).each(($group) => {
    cy.wrap($group).find(listSelector.itemArrow).click()
  })
}

// Get Elements
export const getTreeselct = () => {
  return cy.get(parentSelector)
}

export const getList = () => {
  return cy.get(listSelector.base)
}

export const getInput = () => {
  return cy.get(inputSelector.base)
}

export const getTagsElements = () => {
  return cy.get(inputSelector.tagsElement)
}
