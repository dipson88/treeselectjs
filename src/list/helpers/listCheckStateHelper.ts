import { ValueOptionType, FlattedOptionType } from '../../treeselectTypes'
import { getChildrenOptions } from './listOptionsHelper'

export const updateOptionsByValue = (
  newValue: ValueOptionType[],
  flattedOptions: FlattedOptionType[],
  isSingleSelect: boolean,
  isIndependentNodes: boolean
) => {
  uncheckedAllFlattedOptions(flattedOptions)
  const optionsToCheck = flattedOptions.filter((option) => !option.disabled && newValue.some((id) => id === option.id))

  if (isSingleSelect && !!optionsToCheck.length) {
    optionsToCheck[0].checked = true

    return
  }

  optionsToCheck.forEach((option) => {
    option.checked = true
    // option.isPartialChecked = false
    const resultChecked = updateOptionByCheckState(option, flattedOptions, isIndependentNodes)
    option.checked = resultChecked
  })
}

export const updateOptionByCheckState = (
  { id, checked }: Partial<FlattedOptionType>,
  flattedOptions: FlattedOptionType[],
  isIndependentNodes: boolean
) => {
  const currentOption = flattedOptions.find((option) => option.id === id)

  if (!currentOption) {
    return false
  }

  if (isIndependentNodes) {
    currentOption.checked = currentOption.disabled ? false : !!checked

    return currentOption.checked
  }

  const resultCheckedState = updateFlattedOptionStateWithChildren(!!checked, currentOption, flattedOptions)
  updateParentFlattedOptions(currentOption, flattedOptions)

  return resultCheckedState
}

const updateFlattedOptionStateWithChildren = (
  checked: boolean,
  currentOption: FlattedOptionType,
  flattedOptions: FlattedOptionType[]
) => {
  if (!currentOption.isGroup) {
    currentOption.checked = currentOption.disabled ? false : !!checked
    currentOption.isPartialChecked = false

    return currentOption.checked
  }

  const childrenOptions = flattedOptions.filter((option) => option.childOf === currentOption.id)
  const falseOrDisabledOrPartial = !checked || currentOption.disabled || currentOption.isPartialChecked

  if (falseOrDisabledOrPartial) {
    currentOption.checked = false
    currentOption.isPartialChecked = false
    checkUncheckAllChildren(currentOption, childrenOptions, flattedOptions)

    return currentOption.checked
  }

  const canWeCheckAllChildren = !isSomeChildrenDisabled(childrenOptions, flattedOptions)

  if (canWeCheckAllChildren) {
    currentOption.checked = true
    currentOption.isPartialChecked = false
    checkUncheckAllChildren(currentOption, childrenOptions, flattedOptions)

    return currentOption.checked
  }

  const isAllDisabled = isAllChildrenDisabled(childrenOptions)

  if (isAllDisabled) {
    currentOption.checked = false
    currentOption.isPartialChecked = false
    currentOption.disabled = true

    return currentOption.checked
  }

  currentOption.checked = false
  currentOption.isPartialChecked = true

  childrenOptions.forEach((option) => {
    updateFlattedOptionStateWithChildren(checked, option, flattedOptions)
  })

  return currentOption.checked
}

const updateParentFlattedOptions = (childNode: FlattedOptionType, flattedOptions: FlattedOptionType[]) => {
  const parentOption = flattedOptions.find((option) => option.id === childNode.childOf)

  if (!parentOption) {
    return
  }

  updateParentOption(parentOption, flattedOptions)
  updateParentFlattedOptions(parentOption, flattedOptions)
}

const updateParentOption = (parentOption: FlattedOptionType, flattedOptions: FlattedOptionType[]) => {
  const children = getChildrenOptions(parentOption, flattedOptions)
  const isAllDisabled = isAllChildrenDisabled(children)

  if (isAllDisabled) {
    parentOption.checked = false
    parentOption.isPartialChecked = false
    parentOption.disabled = true

    return
  }

  const isAllChecked = isAllChildrenChecked(children)

  if (isAllChecked) {
    parentOption.checked = true
    parentOption.isPartialChecked = false

    return
  }

  const isSomeCheckedOrPartial = isSomeChildrenCheckedOrPartial(children)

  if (isSomeCheckedOrPartial) {
    parentOption.checked = false
    parentOption.isPartialChecked = true

    return
  }

  parentOption.checked = false
  parentOption.isPartialChecked = false
}

const checkUncheckAllChildren = (
  { checked, disabled }: Partial<FlattedOptionType>,
  children: FlattedOptionType[],
  flattedOptions: FlattedOptionType[]
) => {
  children.forEach((option) => {
    option.disabled = !!disabled || !!option.disabled
    option.checked = !!checked && !option.disabled
    option.isPartialChecked = false
    const subChildren = getChildrenOptions(option, flattedOptions)
    checkUncheckAllChildren({ checked, disabled }, subChildren, flattedOptions)
  })
}

const isSomeChildrenDisabled = (children: FlattedOptionType[], flattedOptions: FlattedOptionType[]) => {
  const isSomeDisabled = children.some((option) => option.disabled)

  if (isSomeDisabled) {
    return true
  }

  const isSomeSubChildrenDisabled = children.some((option) => {
    if (option.isGroup) {
      const subChildren = getChildrenOptions(option, flattedOptions)

      return isSomeChildrenDisabled(subChildren, flattedOptions)
    }

    return false
  })

  return isSomeSubChildrenDisabled
}

const isAllChildrenDisabled = (children: FlattedOptionType[]) => {
  return children.every((option) => !!option.disabled)
}

const isAllChildrenChecked = (children: FlattedOptionType[]) => {
  return children.every((option) => !!option.checked)
}

const isSomeChildrenCheckedOrPartial = (children: FlattedOptionType[]) => {
  return children.some((option) => !!option.checked || !!option.isPartialChecked)
}

const uncheckedAllFlattedOptions = (flattedOptions: FlattedOptionType[]) => {
  flattedOptions.forEach((option) => {
    option.checked = false
    option.isPartialChecked = false
  })
}
