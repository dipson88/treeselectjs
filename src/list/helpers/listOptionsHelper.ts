import { OptionType, ValueOptionType, FlattedOptionType } from '../../treeselectTypes'
import { updateOptionByCheckState } from './listCheckStateHelper'

export const getFlattedOptions = (options: OptionType[], openLevel: number, isIndependentNodes: boolean) => {
  const defaultParams = { level: 0, groupId: '' }
  const flattedOptions = getInitFlattedOptions(options, openLevel, defaultParams.groupId, defaultParams.level)
  const adjustedFlattedOptions = getAdjustedFlattenOptionsUpdate(flattedOptions, isIndependentNodes)

  return adjustedFlattedOptions
}

const getInitFlattedOptions = (options: OptionType[], openLevel: number, groupId: ValueOptionType, level: number) => {
  return options.reduce((acc, curr) => {
    const isGroup = !!curr.children?.length
    const isClosed = level >= openLevel && isGroup
    const hidden = level > openLevel

    acc.push({
      id: curr.value,
      name: curr.name,
      childOf: groupId,
      isGroup,
      checked: false,
      isPartialChecked: false,
      level,
      isClosed,
      hidden,
      disabled: curr.disabled ?? false
    })

    if (isGroup) {
      const children = getInitFlattedOptions(curr.children, openLevel, curr.value, level + 1)
      acc.push(...children)
    }

    return acc
  }, [] as FlattedOptionType[])
}

export const getChildrenOptions = ({ id }: Partial<FlattedOptionType>, flattedOptions: FlattedOptionType[]) => {
  return flattedOptions.filter((option) => option.childOf === id)
}

export const getCheckedOptions = (flattedOptions: FlattedOptionType[]) => {
  const { ungroupedNodes, allGroupedNodes, allNodes } = flattedOptions.reduce(
    (acc, curr) => {
      if (!curr.checked) {
        return acc
      }

      acc.allNodes.push(curr)

      if (curr.isGroup) {
        acc.allGroupedNodes.push(curr)
      } else {
        acc.ungroupedNodes.push(curr)
      }

      return acc
    },
    {
      ungroupedNodes: [] as FlattedOptionType[],
      allGroupedNodes: [] as FlattedOptionType[],
      allNodes: [] as FlattedOptionType[]
    }
  )

  const groupedNodes = allNodes.filter((node) => !allGroupedNodes.some(({ id }) => id === node.childOf))

  return { ungroupedNodes, groupedNodes, allNodes }
}

const getAdjustedFlattenOptionsUpdate = (flattedOptions: FlattedOptionType[], isIndependentNodes: boolean) => {
  // Disabled update
  const disabledNodes = flattedOptions.filter((option) => !!option.disabled)
  disabledNodes.forEach(({ id }) =>
    updateOptionByCheckState({ id, checked: false }, flattedOptions, isIndependentNodes)
  )
  // TODO etc updates

  return flattedOptions
}
