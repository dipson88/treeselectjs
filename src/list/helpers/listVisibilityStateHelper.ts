import { ValueOptionType, FlattedOptionType } from '../../treeselectTypes'
import { getChildrenOptions } from './listOptionsHelper'

export const hideShowChildrenOptions = (
  flattedOptions: FlattedOptionType[],
  { id, isClosed }: Partial<FlattedOptionType>
) => {
  const allChildrenOptions = getChildrenOptions({ id }, flattedOptions)

  allChildrenOptions.forEach((option) => {
    option.hidden = isClosed ?? false

    if (option.isGroup && !option.isClosed) {
      hideShowChildrenOptions(flattedOptions, { id: option.id, isClosed })
    }
  })
}

export const expandSelectedItems = (flattedOptions: FlattedOptionType[]) => {
  flattedOptions
    .filter((option) => option.isGroup && !option.disabled && (option.checked || option.isPartialChecked))
    .forEach((option) => {
      option.isClosed = false
      hideShowChildrenOptions(flattedOptions, option)
    })
}

export const updateVisibleBySearchFlattedOptions = (flattedOptions: FlattedOptionType[], searchText: string) => {
  const optionsWithSearchText = getSearchedFlattedOptions(flattedOptions, searchText)

  flattedOptions.forEach((option) => {
    const isVisible = optionsWithSearchText.some(({ id }) => id === option.id)

    if (isVisible) {
      if (option.isGroup) {
        option.isClosed = false
        hideShowChildrenOptions(flattedOptions, option)
      }

      option.hidden = false
    } else {
      option.hidden = true
    }
  })
}

const getSearchedFlattedOptions = (flattedOptions: FlattedOptionType[], searchText: string) => {
  return flattedOptions.reduce((acc, curr) => {
    const isSearched = curr.name.toLowerCase().includes(searchText.toLowerCase())

    if (isSearched) {
      acc.push(curr)

      if (curr.isGroup) {
        const flattedChildren = getAllNestedChildren(curr.id, flattedOptions)
        acc.push(...flattedChildren)
      }

      if (curr.childOf) {
        const flattedParents = getAllNestedParents(curr.childOf, flattedOptions)
        acc.push(...flattedParents)
      }
    }

    return acc
  }, [] as FlattedOptionType[])
}

const getAllNestedChildren = (childOf: ValueOptionType, flattedOption: FlattedOptionType[]) => {
  return flattedOption.reduce((acc, curr) => {
    if (curr.childOf === childOf) {
      acc.push(curr)

      if (curr.isGroup) {
        acc.push(...getAllNestedChildren(curr.id, flattedOption))
      }
    }

    return acc
  }, [] as FlattedOptionType[])
}

const getAllNestedParents = (childOf: ValueOptionType, flatOptions: FlattedOptionType[]) => {
  return flatOptions.reduce((acc, curr) => {
    if (curr.id === childOf) {
      acc.push(curr)

      if (curr.childOf) {
        acc.push(...getAllNestedParents(curr.childOf, flatOptions))
      }
    }

    return acc
  }, [] as FlattedOptionType[])
}
