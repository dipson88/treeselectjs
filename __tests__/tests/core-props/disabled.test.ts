import { fireEvent } from '@testing-library/dom'
import { renderTreeselect } from '../../helpers'

describe('disabled prop', () => {
  it('should render a disabled Treeselect', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [],
      disabled: true
    })

    expect(treeselect.parentHtmlContainer.classList.contains('treeselect--disabled')).toBe(true)
  })

  it('should not open the dropdown when Treeselect is disabled', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [],
      disabled: true
    })

    const arrow = treeselect.parentHtmlContainer.querySelector('.treeselect-input__arrow') as HTMLElement
    fireEvent.mouseDown(arrow)
    expect(document.body.innerHTML).not.toContain('.treeselect-list')
  })

  it('should not remove tags when Treeselect is disabled', () => {
    const name = 'Option 1'

    const treeselect = renderTreeselect({
      value: [1],
      options: [{ value: 1, name, children: [] }],
      disabled: true
    })

    const tagsElement = treeselect.parentHtmlContainer.querySelector('.treeselect-input__tags') as HTMLElement
    fireEvent.mouseDown(tagsElement)
    expect(tagsElement.innerHTML).toContain(name)
  })

  it('should not be focused when Treeselect is disabled', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [],
      disabled: true
    })

    const input = treeselect.parentHtmlContainer.querySelector('.treeselect-input__edit') as HTMLElement
    expect(input.getAttribute('tabindex')).toBe('-1')
  })

  it('should be focused when Treeselect is enabled', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: []
    })

    const input = treeselect.parentHtmlContainer.querySelector('.treeselect-input__edit') as HTMLElement
    expect(input.getAttribute('tabindex')).not.toBe('-1')
  })
})
