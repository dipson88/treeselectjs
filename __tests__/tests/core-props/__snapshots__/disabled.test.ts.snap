// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disabled prop should not remove tags when Treeselect is disabled 1`] = `
<div
  class="treeselect treeselect--disabled"
  id="test-container"
>
  <div
    class="treeselect-input treeselect-input--bottom treeselect-input--opened"
    tabindex="-1"
  >
    <div
      class="treeselect-input__tags"
    >
      <div
        class="treeselect-input__tags-element"
        tabindex="-1"
        tag-id="1"
        title="Option 1"
      >
        <span
          class="treeselect-input__tags-name"
        >
          Option 1
        </span>
        <span
          class="treeselect-input__tags-cross"
        >
          <svg
            fill="none"
            height="15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 25 25"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </span>
      </div>
      <input
        class="treeselect-input__edit"
        readonly="readonly"
        tabindex="-1"
      />
    </div>
    <div
      class="treeselect-input__operators"
    >
      <span
        class="treeselect-input__clear"
        tabindex="-1"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="15"
            x2="9"
            y1="9"
            y2="15"
          />
          <line
            x1="9"
            x2="15"
            y1="9"
            y2="15"
          />
        </svg>
      </span>
      <span
        class="treeselect-input__arrow"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 15l-6-6-6 6"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="treeselect-list treeselect-list--bottom"
    direction="bottom"
  >
    <div
      class="treeselect-list__item treeselect-list__item--checked treeselect-list__item--focused"
      group="false"
      level="0"
      style="padding-left: 5px;"
      tabindex="-1"
      title="Option 1"
    >
      <div
        class="treeselect-list__item-checkbox-container"
      >
        <span
          class="treeselect-list__item-checkbox-icon"
        >
          <svg
            fill="none"
            height="15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 25 25"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="20 6 9 17 4 12"
            />
          </svg>
        </span>
        <input
          class="treeselect-list__item-checkbox"
          input-id="1"
          tabindex="-1"
          type="checkbox"
        />
      </div>
      <label
        class="treeselect-list__item-label"
      >
        Option 1
      </label>
    </div>
    <div
      class="treeselect-list__empty treeselect-list__empty--hidden"
      title="No results found..."
    >
      <span
        class="treeselect-list__empty-icon"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"
          />
          <line
            x1="12"
            x2="12"
            y1="9"
            y2="13"
          />
          <line
            x1="12"
            x2="12.01"
            y1="17"
            y2="17"
          />
        </svg>
      </span>
      <span
        class="treeselect-list__empty-text"
      >
        No results found...
      </span>
    </div>
  </div>
</div>
`;

exports[`disabled prop should render a disabled Treeselect 1`] = `
<div
  class="treeselect treeselect--disabled"
  id="test-container"
>
  <div
    class="treeselect-input treeselect-input--value-not-selected"
    tabindex="-1"
  >
    <div
      class="treeselect-input__tags"
    >
      <input
        class="treeselect-input__edit"
        placeholder="Search..."
        readonly="readonly"
        tabindex="-1"
      />
    </div>
    <div
      class="treeselect-input__operators"
    >
      <span
        class="treeselect-input__clear"
        tabindex="-1"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="15"
            x2="9"
            y1="9"
            y2="15"
          />
          <line
            x1="9"
            x2="15"
            y1="9"
            y2="15"
          />
        </svg>
      </span>
      <span
        class="treeselect-input__arrow"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 9l6 6 6-6"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
